### DISCUSSION TITLE
Tab completion

### DISCUSSION THREAD

**Discussion by Albert24GG** [LazyVim/LazyVim#250] [Q&A] [answered]
After installing the LazyVim Starter, most things work just fine, except tab completion. Whent typing the code, suggestions show up, but when I press tab, it just inserts an indentation block . Also, having the starter config, all the configs from the docs are set https://www.lazyvim.org/configuration/recipes#supertab, including the luasnip and nvim-cmp config. How can I make this work, since it is annoying to have to scroll through the completions list every time?

**Comment by folke** [comment]
I'm sorry, but why do you keep bringing this up? I assume you are the same person that posted it twice on Reddit, opened an issue and now here?

Hundreds of users use the supertab recipe without any issues. I even took the time, to retest this myself and it all works as expected with a clean LazyVim install. 

You clearly have something else in your config that is messing things up. Please debug your own config.

You can use things like `:verbose imap <tab>` to figure out what's wrong.

**Comment by folke** [comment, marked as answer]
Are you sure that the supertab config is even loaded? Add a print statement in the `opts` function to verify. If it doens't print anything, then you added the config in the wrong file.

**Reply by smokeyhallow** [reply]
Can you explain where to add the print to debug?

I added it as follows:

``` lua
-- ...
  {
    "hrsh7th/nvim-cmp",
    dependencies = {
      "hrsh7th/cmp-emoji",
    },
    ---@param opts cmp.ConfigSchema
    opts = function(_, opts)
      print("Hello from supertab")
      local has_words_before = function()
        unpack = unpack or table.unpack
        local line, col = unpack(vim.api.nvim_win_get_cursor(0))
        return col ~= 0 and vim.api.nvim_buf_get_lines(0, line - 1, line, true)[1]:sub(col, col):match("%s") == nil
      end

      local luasnip = require("luasnip")

-- ...
```

Is that where you meant? And where can I find the output printed?

**Reply by zoyron** [reply]
Thank you so much for posting the recipe. I love lazyvim config, but not being able to use tab and s-tab was bothering me. This works just as I want

**Comment by folke** [comment]
Creating multiple reddit posts, issues and now this is just not cool.

**Reply by Albert24GG** [reply]
I'm sorry if I bothered, but I'm not the person from reddit. I just wanted to solve this issue since I genuinely like this nvim config. I managed to make it work. The problem was that the plugins config wasn't actually returning anything. Debugging the way you suggested helped. Thank you!

**Reply by folke** [reply]
ok, great. Glad you got it working!

**Reply by smokeyhallow** [reply]
@Albert24GG can you confirm whether your requirement was to confirm/select the item from the drop down using tab, than enter?

Also can you confirm whether you were able to solve with the steps mentioned in [SuperTab section](https://www.lazyvim.org/configuration/recipes#supertab)?

**Reply by Albert24GG** [reply]
@smokeyhallow I only managed to use tab for scrolling through the list(with the config mentioned in that link), but I have to press enter to confirm my selection.

**Comment by smokeyhallow** [comment]
I am very new to lua & lazyvim. @folke I know you are very frustrated with the same question again & again. But, I too can't understand how to solve this.

This is what I did:

Created a file `supertab.lua` in `./nvim/lua/plugins/`

With the following contents copied from [#supertab](https://www.lazyvim.org/configuration/recipes#supertab)

``` lua
return {
{
  "L3MON4D3/LuaSnip",
  keys = function()
    return {}
  end,
},
{
  "hrsh7th/nvim-cmp",
  dependencies = {
    "hrsh7th/cmp-emoji",
  },
  ---@param opts cmp.ConfigSchema
  opts = function(_, opts)
    local has_words_before = function()
      unpack = unpack or table.unpack
      local line, col = unpack(vim.api.nvim_win_get_cursor(0))
      return col ~= 0 and vim.api.nvim_buf_get_lines(0, line - 1, line, true)[1]:sub(col, col):match("%s") == nil
    end

    local luasnip = require("luasnip")
    local cmp = require("cmp")

    opts.mapping = vim.tbl_extend("force", opts.mapping, {
      ["<Tab>"] = cmp.mapping(function(fallback)
        if cmp.visible() then
          cmp.select_next_item()
        -- You could replace the expand_or_jumpable() calls with expand_or_locally_jumpable()
        -- they way you will only jump inside the snippet region
        elseif luasnip.expand_or_jumpable() then
          luasnip.expand_or_jump()
        elseif has_words_before() then
          cmp.complete()
        else
          fallback()
        end
      end, { "i", "s" }),
      ["<S-Tab>"] = cmp.mapping(function(fallback)
        if cmp.visible() then
          cmp.select_prev_item()
        elseif luasnip.jumpable(-1) then
          luasnip.jump(-1)
        else
          fallback()
        end
      end, { "i", "s" }),
    })
  end,
}
}
```

Is this the recommended way?

**Reply by folke** [reply]
that looks correct. what isn't working specifically?

**Reply by smokeyhallow** [reply]
When items are populated in popup, pressing tab cycles through the list than completing the selected option. Only pressing enter works.

I trust the above snippet is supposed to bring that behavior?

**Reply by folke** [reply]
I have no idea. That's the snippet from the cmp wiki. I don't use it myself. I've added it because a lot of people were asking for it.

I think it's to use `tab` and `s-tab` to cycle though the items in the completion list, so you'd still have to use `cr` or `c-e` to confirm a completion.

**Reply by smokeyhallow** [reply]
Thanks, I understand this isn't something you are using. I will find a way, but can you confirm:

- Is it technically possible to have the completion be trigger via `tab` than `enter`?
- And prevent the item cycling behaviour thats implemented now?

I find this a huge cognitive dissonance to having to press enter to confirm a selection. I am so used to tab that it trips every time. Thats why I am desperately trying for a fix.

**Comment by smokeyhallow** [comment]
Figured this out!!

Found it from `nvim-cmp`'s [wiki](https://github.com/hrsh7th/nvim-cmp/wiki/Example-mappings#intellij-like-mapping).

This enables tab to complete feature for Lazy.vim. I guess this may be of help for many soon.

``` lua
return {
  {
    "L3MON4D3/LuaSnip",
    keys = function()
      return {}
    end,
  },
  {
    "hrsh7th/nvim-cmp",
    dependencies = {
      "hrsh7th/cmp-emoji",
    },
    ---@param opts cmp.ConfigSchema
    opts = function(_, opts)
      local cmp = require("cmp")

      opts.mapping = vim.tbl_extend("force", opts.mapping, {
        ["<Tab>"] = cmp.mapping(function(fallback)
          -- This little snippet will confirm with tab, and if no entry is selected, will confirm the first item
          if cmp.visible() then
            local entry = cmp.get_selected_entry()
            if not entry then
              cmp.select_next_item({ behavior = cmp.SelectBehavior.Select })
            else
              cmp.confirm()
            end
          else
            fallback()
          end
        end, { "i", "s", "c" }),
      })
    end,
  },
}

```

**Reply by smokeyhallow** [reply]
@folke I think this should be part of the recipes

**Reply by folke** [reply]
That's exactly what's in the supertab recipe on the website. So shy should this be on it?

**Reply by temefleet** [reply]
This is the behaviour I was looking for too. It's not the same as in the recipe. This actually selects the first option with tab.

**Reply by ThatIsCraZy** [reply]
one year later and the recipe still contains the wrong code.
the code from the supertab recipe, changes Nothing of the default behaviour . the script mention by @smokeyhallow  does the trick.

**Reply by nipuna-perera** [reply]
This is the expected behavior for me also. The recipe in the docs actually doesn't change anything.

**Reply by entropy36** [reply]
hello, I'm totally new to lazyvim, I want to know in which file should this be added?

**Reply by entropy36** [reply]
> hello, I'm totally new to lazyvim, I want to know in which file should this be added?

oh，I got it. But here is another problem, when suggestions show up and I press Tab, it just select the first suggesion and i can't switch between them. What should I do?

**Comment by metala** [comment]
If you want to prevent `<CR>` (enter key) to complete the suggestion add following to the overridden `opts.mapping` to the smokeyhallow's solution:
```lua
        ["<CR>"] = cmp.mapping(function(fallback)
          cmp.abort()
          fallback()
        end, { "i", "s", "c" }),
```

**Reply by meleu-cw** [reply]
Hey @metala, could you explain in which file this content should be placed?

**Reply by metala** [reply]
Oh, sorry. I wasn't paying attention to this platform.

It's the `opts` argument to the `hrsh7th/nvim-cmp` plugin.
Here is the part from my config at `.config/nvim/lua/plugins/user.lua`. However, I'd be dropping LazyVim and I will switch to a more lean setup.
```lua
  {
    "hrsh7th/nvim-cmp",
    dependencies = {
      "hrsh7th/cmp-emoji",
    },
    ---@param opts cmp.ConfigSchema
    opts = function(_, opts)
      local cmp = require("cmp")

      opts.mapping = vim.tbl_extend("force", opts.mapping, {
        ["<Tab>"] = cmp.mapping(function(fallback)
          -- This little snippet will confirm with tab, and if no entry is selected, will confirm the first item
          if cmp.visible() then
            local entry = cmp.get_selected_entry()
            if not entry then
              cmp.select_next_item({ behavior = cmp.SelectBehavior.Select })
            else
              cmp.confirm()
            end
          else
            fallback()
          end
        end, { "i", "s", "c" }),
        ["<CR>"] = cmp.mapping(function(fallback)
          cmp.abort()
          fallback()
        end, { "i", "s", "c" }),
      })
    end,
  },
```

**Comment by macintacos** [comment]
[I'm writing this comment because it seems like there's some confusion going on](https://xkcd.com/386/), and it seems that some folks here understand what `supertab` is and some don't. I think it's worth clarifying that folks who are finding this issue are probably trying to choose one of two options:

* **Option 1:** "I press `tab` and the item in the autocomplete menu is accepted and I can start writing more code."
* **Option 2:** "I want to replicate `supertab` plugin functionality using `nvim-cmp`."

These are similar concepts in that `tab` is associated with the completion menu, but they are not the same thing. `supertab` is originally a vim plugin that's still actively maintained, check it out here if you'd like: https://github.com/ervandew/supertab

To explain the difference, I think it helps to give a demonstration. I am going to use regular `vim` to show you how supertab is supposed to work.

I put together a basic `.vimrc`, using [`coc.nvim`](https://github.com/neoclide/coc.nvim) for some basic completion options while typing (for demo purposes) and [`vim-plug`](https://github.com/junegunn/vim-plug) for plugin management (it's been a while since I've used vanilla `vim`, but this was how I used to do things). In this video, you can see what the default behavior is of pressing `tab` in `vim` when you have completion options available:

https://github.com/LazyVim/LazyVim/assets/8128941/a2e388e5-73cd-4aa3-933a-e5ad08a582de

As you can see, `tab` doesn't do much of anything except dismiss the completions and insert a literal `tab`. I think everyone in this discussion can probably agree, this isn't what you want. But if what you want is Option 1, that isn't quite how `supertab` works. Installing the `supertab` plugin _does not_ do "tab completion" in the sense that it completes the first item and moves on - rather, `tab` is used to navigate through the list of options, as demonstrated in this video:

https://github.com/LazyVim/LazyVim/assets/8128941/3bd6b5c5-2c03-492a-ab6a-fac5b671afe9

In this video, you can see what `supertab` has done. Pressing `tab` now cycles through the available options in the completion menu, and does not dismiss the completion window at all.

I tried the solution provided [here](https://github.com/LazyVim/LazyVim/discussions/250#discussioncomment-5270817), and can definitively say that this is _not_ `supertab`-like autocompletion. It is Option 1. This is perfectly fine - some people like this kind of completion style. I personally do not, and prefer `supertab`. But it's important to understand that these are two different ways of using the `tab` key, and I think that the recipe in LazyVim already closely maps how I expect `supertab` to function, so there is nothing wrong with the documentation AFAICT.

Now, should there be another entry in the documentation to reference Option 1's style of using `tab` for completion? Should the LazyVim documentation clarify what `supertab` is, and perhaps link to the `supertab` repo so that folks understand what it is? Maybe! But I would not remove the current recipe, because for users like myself who _want_ `supertab`-like functionality, this recipe works perfectly.

Anyway, I hope this helps clear things up! It might be worth putting this in the wiki or something, but I'll leave that up to the maintainers.

**Reply by macintacos** [reply]
Also, just to clarify - I believe that Option 1 would probably be considered more "default VSCode-like tab-completion", which I imagine is what a lot of people coming from VSCode are looking to replicate because they are used to it. Perfectly valid approach, but it still isn't `supertab` :)

**Reply by gitsang** [reply]
Thank you very much, I searched almost all over the web and found that everyone had a completely different understanding of tab complete, which made me miserable because when I tried other people's configurations, I found that I couldn't achieve the results I wanted. I now finally understand that what I need is supertab-like completion, but I still don't know what to do.

I noticed that when you mention `think that the recipe in LazyVim already closely maps how I expect supertab to function, so there is nothing wrong with the documentation AFAICT.`, Do you mean https://www.lazyvim.org/configuration/recipes#supertab?

I tried the configuration of this document but it didn't work, when I clicked on `tab`, it typed in a 'tab' character, which seems to be the manifestation of option 1. Or maybe I'm having trouble configuring? How should I debug?

**Reply by gitsang** [reply]
Ok, I found that nvim-cmp was replaced by blink, https://www.lazyvim.org/news#removed-plugins. And lazyvim seems to prevent users from uninstalling blink.

**Reply by dpetka2001** [reply]
> And lazyvim seems to prevent users from uninstalling blink.

That's just wrong. You just have to enable the `coding.nvim-cmp` Extra via `:LazyExtras` command. Then, `nvim-cmp` will get installed and `blink.cmp` uninstalled.

**Reply by gitsang** [reply]
Thank you, I tried for a long time, it turns out that I just need to enable `coding.nvim-cmp`

**Comment by noprobelm** [comment]
If you are using `blink` for auto completion, you can enable super-tab completion with this snippet from the [documentation](https://cmp.saghen.dev/installation):

```lua
return {
  "saghen/blink.cmp",
  -- optional: provides snippets for the snippet source
  dependencies = "rafamadriz/friendly-snippets",

  -- use a release tag to download pre-built binaries
  version = "*",
  -- AND/OR build from source, requires nightly: https://rust-lang.github.io/rustup/concepts/channels.html#working-with-nightly-rust
  -- build = 'cargo build --release',
  -- If you use nix, you can build from source using latest nightly rust with:
  -- build = 'nix run .#build-plugin',

  ---@module 'blink.cmp'
  ---@type blink.cmp.Config
  opts = {
    -- 'default' for mappings similar to built-in completion
    -- 'super-tab' for mappings similar to vscode (tab to accept, arrow keys to navigate)
    -- 'enter' for mappings similar to 'super-tab' but with 'enter' to accept
    -- See the full "keymap" documentation for information on defining your own keymap.
    keymap = { preset = "super-tab" },

    appearance = {
      -- Sets the fallback highlight groups to nvim-cmp's highlight groups
      -- Useful for when your theme doesn't support blink.cmp
      -- Will be removed in a future release
      use_nvim_cmp_as_default = true,
      -- Set to 'mono' for 'Nerd Font Mono' or 'normal' for 'Nerd Font'
      -- Adjusts spacing to ensure icons are aligned
      nerd_font_variant = "mono",
    },

    -- Default list of enabled providers defined so that you can extend it
    -- elsewhere in your config, without redefining it, due to `opts_extend`
    sources = {
      default = { "lsp", "path", "snippets", "buffer" },
    },
  },
  opts_extend = { "sources.default" },
}
```

**Reply by gitsang** [reply]
If you want to implement the Option 2 described by [macintacos](https://github.com/macintacos)

```lua
    keymap = {
      preset = "default",
      ["<Tab>"] = { "select_next", "fallback" },
      ["<S-Tab>"] = { "select_prev", "fallback" },
    },
```

**Reply by errkk** [reply]
Looks like blink is suddenly now the default, rather than nvim-cmp, despite the recipe for super-tab in the docs still describing the way to do it for nvim-cmp. so took me a while get to get it to behave as my muscle memory expected. but this seems to do it!

**Reply by thomas-pfeil** [reply]
So for beginners like me the full solution:

- install `coding.blink` via `Lazy Extras`
- add a plugin file `~/.config/nvim/lua/plugins/blink.lua` with content below
```
return {
  "saghen/blink.cmp",
  opts = {
    keymap = {
      preset = "default",
      ["<Tab>"] = { "select_next", "fallback" },
      ["<S-Tab>"] = { "select_prev", "fallback" },
      ['<CR>'] = { 'accept', 'fallback' },
    },
  },
}
```

**Reply by codingedward** [reply]
If you also want the first item on the list be selected the first time you press Tab, you can use this setting:

```
  return {
    "saghen/blink.cmp",
    opts = {
      completion = { list = { selection = { preselect = false, auto_insert = true } }, },
      keymap = {
        preset = "default",
        ["<Tab>"] = { "select_next", "fallback" },
        ["<S-Tab>"] = { "select_prev", "fallback" },
        -- ["<CR>"] = { "accept", "fallback" },
      },
    },
  },

**Reply by baipaiha** [reply]
Thank you !

To avoid a lot of noise on this subject, i think this solution should be added to the `Recipes` documentation with a title like `If you want to cycle with Tab and accept with Enter`
